counterparty-core/counterpartycore/lib/messages/utxo.py

```diff
@@ -109,44 +129,107 @@ def compose(db, source, destination, asset, quantity):
-    elif not destination:  # attach to utxo
+    if not util.is_utxo_format(source) and not destination:  # attach to utxo
         # if no destination, we use the source address as the destination
-        destinations.append((source_address, None))
+        destinations.append((source, None))
 
-    return (source_address, destinations, data)
+    return (source, destinations, data)
 
 
 def unpack(message, return_dict=False):
     try:
         data_content = struct.unpack(f">{len(message)}s", message)[0].decode("utf-8").split("|")
-        (source, destination_str, asset, quantity) = data_content
-        destination = None if destination_str == "" else destination_str
+        (source, destination, asset, quantity) = data_content
+        destination = destination or None
+        asset = asset or None
+        quantity = int(quantity) if quantity else None
 
         if return_dict:
             return {
                 "source": source,
                 "destination": destination,
                 "asset": asset,
-                "quantity": int(quantity),
+                "quantity": quantity,
             }
 
-        return (source, destination, asset, int(quantity))
+        return (source, destination, asset, quantity)
     except Exception as e:
         raise exceptions.UnpackError(f"Cannot unpack utxo message: {e}") from e
 
 
+def pay_fee(db, tx, action, source, recipient):
+    # fee payment only for attach to utxo
+    if action == "attach to utxo":
+        fee = gas.get_transaction_fee(db, ID, tx["block_index"])
+    else:
+        fee = 0
+    if fee > 0:
+        # fee is always paid by the address
+        if action == "attach to utxo":
+            fee_payer = source
+        else:
+            fee_payer = recipient
+        # debit fee from the fee payer
+        ledger.debit(
+            db,
+            fee_payer,
+            config.XCP,
+            fee,
+            tx["tx_index"],
+            action=f"{action} fee",
+            event=tx["tx_hash"],
+        )
+        # destroy fee
+        destroy_bindings = {
+            "tx_index": tx["tx_index"],
+            "tx_hash": tx["tx_hash"],
+            "block_index": tx["block_index"],
+            "source": tx["source"],
+            "asset": config.XCP,
+            "quantity": fee,
+            "tag": f"{action} fee",
+            "status": "valid",
+        }
+        ledger.insert_record(db, "destructions", destroy_bindings, "ASSET_DESTRUCTION")
+    return fee
+
+
+def proceed_send(db, tx, action, source, recipient, asset, quantity, event, fee_paid):
+    # debit asset from source and credit to recipient
+    ledger.debit(db, source, asset, quantity, tx["tx_index"], action=action, event=tx["tx_hash"])
+    ledger.credit(
+        db,
+        recipient,
+        asset,
+        quantity,
+        tx["tx_index"],
+        action=action,
+        event=tx["tx_hash"],
+    )
+    # we store parameters only if the transaction is valid
```

Comment:
where is the validity determined? in `ledger.insert_record()`?

Reply:
I remove this comment from here. Come from a copy/past and it's not relevant here.